In [1]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
import numpy as np
from xgboost import XGBClassifier
from sklearn.linear_model import SGDClassifier
import time
from sklearn.svm import LinearSVC
from sklearn.neural_network import MLPClassifier
warnings.filterwarnings('ignore')

In [2]:
data=pd.read_csv('data.csv')
data.head()

target  ps_ind_01  ps_ind_03  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
0       0          2          5              0              1              0   
1       0          1          7              0              0              1   
2       0          5          9              0              0              1   
3       0          0          2              1              0              0   
4       0          0          0              1              0              0   

   ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  ...  \
0              0              0              0              0  ...   
1              0              0              0              0  ...   
2              0              0              0              0  ...   
3              0              0              0              0  ...   
4              0              0              0              0  ...   

   ps_car_11_cat_104         0         1         2         3         4  \
0                  0  5.048113  2.281937 -0.006126 -1.146353  1.506130   
1                  0 -2.917107  1.109295  2.435316  1.171819 -0.013408   
2                  0  0.404935 -1.454792  1.516882  1.297032  0.161706   
3                  1 -1.160598 -0.545594 -1.213880  0.475253  1.026656   
4                  0  0.849476 -0.703857 -0.210735 -1.249064 -0.840705   

          5         6         7         8  
0  0.523855  1.736757  1.067024  0.401926  
1  0.683728 -2.783007 -0.198828  0.478956  
2  1.158127  0.116747  1.063657 -0.532891  
3 -0.227338 -0.033361  0.935005 -0.642617  
4 -0.110160  0.293483 -1.091242 -0.317573  

[5 rows x 219 columns]

In [3]:
y_data=data['target']
x_data=data.drop(['target'], axis=1)

In [4]:
x_train, x_test, y_train, y_test= train_test_split(x_data, y_data, test_size=0.30, random_state=0)

In [5]:
print(f'Shape of x_train = {x_train.shape} and the shape of y_train = {y_train.shape}')
print(f'Shape of x_test = {x_test.shape} and the shape of y_test = {y_test.shape}')

Shape of x_train = (802925, 218) and the shape of y_train = (802925,)
Shape of x_test = (344111, 218) and the shape of y_test = (344111,)


In [6]:
# creating function to calculate precision recall and accuracy
def conf_mat(y_act,y_pre):
    t0, f1, f0, t1= confusion_matrix(y_act, y_pre).ravel()
    accuracy=(t0+t1)/(t0+t1+f0+f1)
    precision=(t1)/(t1+f1)
    recall=(t1)/(t1+t0)
    f1_score=(2*precision*recall)/(precision+recall)
    return np.round((accuracy*100, recall*100, precision*100, f1_score*100,t0, f1, f0, t1),2) 

# 1. The Simple Logistic Regression Model seems to have high accuracy. Is that what we need at all? What is the problem with this model?

In [7]:
start_time=time.time()
lgr=LogisticRegression()
lgr.fit(x_train, y_train)
lgr_predicted=lgr.predict(x_test)
lgr_acc, lgr_recall, lgr_preci, lgr_f1,t0, f1, f0, t1=conf_mat(y_test, lgr_predicted)
tt_time=round(time.time()-start_time,2)
mat_lgr=('Logistic_regression', lgr_acc, lgr_recall, lgr_preci, lgr_f1,t0, f1, f0, t1, tt_time)
print(f'Accuracy of the logistic regression = {lgr_acc}%\n')
print(f'Recall of the logistic regression = {lgr_recall}%\n')
print(f'Precision of the logistic regression = {lgr_preci}%\n')
print(f'F1 score of the logistic regression = {lgr_f1}%\n')
print(f'Time to train {tt_time} s \n')


Accuracy of the logistic regression = 98.98%

Recall of the logistic regression = 49.72%

Precision of the logistic regression = 99.62%

F1 score of the logistic regression = 66.34%

Time to train 24.42 s 



### Though the accuracy is good, the model have low recall value and here our priority is to find the more number of people who wil file a claim, so the recall should be high and it seems that the current value of recall id 49.68


# 2. Why do you think f1-score is 0.0?

## unable to comment on this question as the f1 score for logistics regression we found is 66.3 and that's no where near 0. 

# 3. What is the precision and recall score for the model?

## Precision score = 99.63 %
## Recall score = 49.68 %

# 4. What is the most important inference you can draw from the result?


##  model is able to recall about 49.5 percent of the data as this is like predicting only the 50 percent of those who can claim the insurance. This method is not preferrable because the positive output have 50 percent chances to be incorrect

# 5. What is the accuracy score and f1-score for the improved Logistic Regression model?

## we can try to use logistic regression with penalty =l2 and solver =sag as sag is used for the large data sets.

##  SGDclassifier with loss function work as a logistic regression

In [8]:
'''solver =sag is used for large data sets'''

lgr=LogisticRegression(penalty='l2', solver='sag')
lgr.fit(x_train, y_train)
lgr_predicted=lgr.predict(x_test)
lgr_acc, lgr_recall, lgr_preci, lgr_f1,t0, f1, f0, t1=conf_mat(y_test, lgr_predicted)
tt_time=round(time.time()-start_time,2)

mat_lgr_sag=('Log Reg with sag',lgr_acc, lgr_recall, lgr_preci, lgr_f1,t0, f1, f0, t1, tt_time)
print(f'Accuracy of the logistic regression = {lgr_acc}%\n')
print(f'Recall of the logistic regression = {lgr_recall}%\n')
print(f'Precision of the logistic regression = {lgr_preci}%\n')
print(f'F1 score of the logistic regression = {lgr_f1}%\n')
print(f'Time to train {tt_time} s \n')


Accuracy of the logistic regression = 99.19%

Recall of the logistic regression = 49.69%

Precision of the logistic regression = 99.9%

F1 score of the logistic regression = 66.37%

Time to train 179.8 s 



In [9]:
'''SGDclassifier with loss function work as a logistic regression'''
start_time=time.time()
lgr=SGDClassifier(loss='log', penalty='l2')
lgr.fit(x_train,y_train)
lgr_predicted=lgr.predict(x_test)
lgr_acc, lgr_recall, lgr_preci, lgr_f1,t0, f1, f0, t1=conf_mat(y_test, lgr_predicted)
tt_time=round(time.time()-start_time,2)
mat_sgd=('SGD classifier', lgr_acc, lgr_recall, lgr_preci, lgr_f1,t0, f1, f0, t1, tt_time)
print(f'Accuracy of the SGDclassifier = {lgr_acc}%\n')
print(f'Recall of the SGDclassifier = {lgr_recall}%\n')
print(f'Precision of the SGDclassifier = {lgr_preci}%\n')
print(f'F1 score of the SGDclassifier = {lgr_f1}%\n')
print(f'Time to train {tt_time} s \n')


Accuracy of the SGDclassifier = 97.59%

Recall of the SGDclassifier = 50.17%

Precision of the SGDclassifier = 97.36%

F1 score of the SGDclassifier = 66.22%

Time to train 21.5 s 



## For SGDclassifier(logistic regression as loss=log) we see that accuracy=97.7% and F1 score = 66.19%

# 6. Why do you think f1-score has improved?

## it doesn't look like there is any significant change in the F1 score, however it can be observed that the accuracy has a negligible decline and the there is a negligible rise on the recall

# 7. For model LinearSVC play with parameters – dual, max_iter and see if there is any improvement

In [10]:
lin_svc=LinearSVC()
start_time=time.time()
lin_svc.fit(x_train, y_train)
lin_svc_predict=lin_svc.predict(x_test)
svc_acc, svc_recall, svc_preci, svc_f1,t0, f1, f0, t1=conf_mat(y_test, lin_svc_predict)
tt_time=round(time.time()-start_time,2)
mat_svc=('SVC', svc_acc, svc_recall, svc_preci, svc_f1,t0, f1, f0, t1, tt_time)
print(f'Accuracy of the LinearSVC = {svc_acc}%\n')
print(f'Recall of the LinearSVC = {svc_recall}%\n')
print(f'Precision of the LinearSVC = {svc_preci}%\n')
print(f'F1 score of the LinearSVC = {svc_f1}%\n')
print(f'Time to train {tt_time} s \n')


Accuracy of the LinearSVC = 99.17%

Recall of the LinearSVC = 49.67%

Precision of the LinearSVC = 99.91%

F1 score of the LinearSVC = 66.36%

Time to train 81.46 s 



In [11]:
'''
max_iter=[500, 1000, 1500, 2000, 2500, 3000]
for j in max_iter:
    print(f'dual = True and max_iter = {j} \n')
    model=LinearSVC(max_iter=j)
    model.fit(x_train,y_train)
    lin_svc_predict=lin_svc.predict(x_test)
    svc_acc, svc_recall, svc_preci, svc_f1=conf_mat(y_test, lin_svc_predict)
    print(f'Accuracy of the LinearSVC = {svc_acc}%\n')
    print(f'Recall of the LinearSVC = {svc_recall}%\n')
    print(f'Precision of the LinearSVC = {svc_preci}%\n')
    print(f'F1 score of the LinearSVC = {svc_f1}%\n')
    print('-X-X'*30)

'''

"\nmax_iter=[500, 1000, 1500, 2000, 2500, 3000]\nfor j in max_iter:\n    print(f'dual = True and max_iter = {j} \n')\n    model=LinearSVC(max_iter=j)\n    model.fit(x_train,y_train)\n    lin_svc_predict=lin_svc.predict(x_test)\n    svc_acc, svc_recall, svc_preci, svc_f1=conf_mat(y_test, lin_svc_predict)\n    print(f'Accuracy of the LinearSVC = {svc_acc}%\n')\n    print(f'Recall of the LinearSVC = {svc_recall}%\n')\n    print(f'Precision of the LinearSVC = {svc_preci}%\n')\n    print(f'F1 score of the LinearSVC = {svc_f1}%\n')\n    print('-X-X'*30)\n\n"

## No significant change was seen in the performance for different values of iteration and dual = True


In [12]:
"""
max_iter=[100, 200, 300, 400,500 ]
for j in max_iter:
    print(f'dual = bool and max_iter = {j} \n')
    model=LinearSVC(dual=bool, max_iter=j, loss='hinge', penalty='l2')
    model.fit(x_train,y_train)
    lin_svc_predict=lin_svc.predict(x_test)
    svc_acc, svc_recall, svc_preci, svc_f1=conf_mat(y_test, lin_svc_predict)
    print(f'Accuracy of the LinearSVC = {svc_acc}%\n')
    print(f'Recall of the LinearSVC = {svc_recall}%\n')
    print(f'Precision of the LinearSVC = {svc_preci}%\n')
    print(f'F1 score of the LinearSVC = {svc_f1}%\n')
    print('-X-X'*30)"""
### using dual =bool gives an error for l1 and l2 combination with hinge and squared hinge as n_samples> n_features
'''
ValueError: Unsupported set of arguments: The combination of penalty='l2' 
and loss='hinge' are not supported when dual=<class 'bool'>, 
Parameters: penalty='l2', loss='hinge', dual=<class 'bool'>
'''

"\nValueError: Unsupported set of arguments: The combination of penalty='l2' \nand loss='hinge' are not supported when dual=<class 'bool'>, \nParameters: penalty='l2', loss='hinge', dual=<class 'bool'>\n"

# 8. SVC with Imbalance Check & Feature Optimization & only 100K Records → is there improvement in scores?

In [13]:
'''
https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html
as per the reference it's advised to use LinearSVC above 10k, however SVC is used here to check what happen's for 100k records
'''
'''taking random 100k records'''
'''data set has been balanced and optimized previously'''

data_svc=pd.read_csv('data.csv')
data_svc=data_svc.sample(n=100000, random_state=1)
y_svc=data_svc['target']
x_svc=data_svc.drop(['target'], axis=1)
xsvc_train, xsvc_test, ysvc_train, ysvc_test= train_test_split(x_svc, y_svc, test_size=0.30, random_state=0)
from sklearn.svm import SVC
model_svc=SVC()

start_time=time.time()
model_svc.fit(xsvc_train, ysvc_train)
ssvc_predict=model_svc.predict(xsvc_test)

ssvc_acc, ssvc_recall, ssvc_preci, ssvc_f1,t0, f1, f0, t1=conf_mat(ysvc_test, ssvc_predict)
tt_time=round(time.time()-start_time,2)
mat_xscv=('SVC with 100',ssvc_acc, ssvc_recall, ssvc_preci, ssvc_f1,t0, f1, f0, t1, tt_time)
print(f'Accuracy of the SVC = {svc_acc}%\n')
print(f'Recall of the SVC = {svc_recall}%\n')
print(f'Precision of the SVC = {svc_preci}%\n')
print(f'F1 score of the SVC = {svc_f1}%\n')
print(f'Time to train {tt_time} s \n')



Accuracy of the SVC = 99.17%

Recall of the SVC = 49.67%

Precision of the SVC = 99.91%

F1 score of the SVC = 66.36%

Time to train 249.38 s 



## It doesn't look like the model has imporved as the recall and f1 is all the same.

# 9. XGBoost is one the better classifiers -- but still f1-score is very low. What could be the reason?

In [14]:
xgb_class=XGBClassifier()
start_time=time.time()
xgb_class.fit(x_train, y_train)
xgb_predict=xgb_class.predict(x_test)
xgb_acc, xgb_recall, xgb_preci, xgb_f1,t0, f1, f0, t1=conf_mat(y_test, xgb_predict)
tt_time=round(time.time()-start_time,2)
mat_xgb=('XGBclassifier',xgb_acc, xgb_recall, xgb_preci, xgb_f1, t0, f1, f0, t1,tt_time)
print(f'Accuracy of the xgb = {xgb_acc}%\n')
print(f'Recall of the xgb = {xgb_recall}%\n')
print(f'Precision of the xgb = {xgb_preci}%\n')
print(f'F1 score of the xgb = {xgb_f1}%\n')
print(f'Time to train {tt_time} s \n')


Accuracy of the xgb = 97.71%

Recall of the xgb = 49.96%

Precision of the xgb = 97.89%

F1 score of the xgb = 66.16%

Time to train 274.8 s 



# 10. What is the increase in number of features after one-hot encoding of the data?

## in Q16 of EDA, we used get_dummies rather than onehotencoding for categorical values.
## There are 14 catergorical features in the original dataset
## after getting dummies and dropping first column we had 177 features
## Total features =177+14=191
## increase in feature after onehot encoding would be = 191-14= 177

# 14. Do you think using AdaBoost can give any significant improvement over XGBoost?

In [15]:
abc_class=AdaBoostClassifier(n_estimators=30, learning_rate=.01, random_state=0)
start_time=time.time()
abc_class.fit(x_train, y_train)
abc_predict=abc_class.predict(x_test)
tt_time=round(time.time()-start_time,2)
abc_acc, abc_recall, abc_preci, abc_f1,t0, f1, f0, t1=conf_mat(y_test, abc_predict)
tt_time=round(time.time()-start_time,2)
mat_ada=('ada_boost',abc_acc, abc_recall, abc_preci, abc_f1, t0, f1, f0, t1,tt_time)
print(f'Accuracy of the abc = {abc_acc}%\n')
print(f'Recall of the abc = {abc_recall}%\n')
print(f'Precision of the abc = {abc_preci}%\n')
print(f'F1 score of the abc = {abc_f1}%\n')
print(f'Time to train {tt_time} s \n')


Accuracy of the abc = 84.75%

Recall of the abc = 59.05%

Precision of the abc = 76.65%

F1 score of the abc = 66.71%

Time to train 113.21 s 



# 15. MLPClassifier is the neural network we are trying. But how to choose the right no. of layers and size?

In [16]:
mlp_model=MLPClassifier(hidden_layer_sizes=(171,100,50), max_iter=700, alpha=0.0001,
                       solver='adam', verbose=2, random_state=0, tol=0.0001)
start_time=time.time()
mlp_model.fit(x_train, y_train)
tt_time=round(time.time()-start_time,2)
mlp_predict=mlp_model.predict(x_test)
mlp_acc, mlp_recall, mlp_preci, mlp_f1,t0, f1, f0, t1=conf_mat(y_test, mlp_predict)
tt_time=round(time.time()-start_time,2)
mat_mlp=('MLPclassifier',mlp_acc, mlp_recall, mlp_preci, mlp_f1,t0, f1, f0, t1, tt_time)
print(f'Accuracy of the mlp = {mlp_acc}%\n')
print(f'Recall of the mlp = {mlp_recall}%\n')
print(f'Precision of the mlp = {mlp_preci}%\n')
print(f'F1 score of the mlp = {mlp_f1}%\n')
print(f'Time to train {tt_time} s \n')


Iteration 1, loss = 0.04518623
Iteration 2, loss = 0.03325763
Iteration 3, loss = 0.03161966
Iteration 4, loss = 0.03084302
Iteration 5, loss = 0.03003295
Iteration 6, loss = 0.02972688
Iteration 7, loss = 0.02910629
Iteration 8, loss = 0.02885735
Iteration 9, loss = 0.02846786
Iteration 10, loss = 0.02824646
Iteration 11, loss = 0.02792817
Iteration 12, loss = 0.02775162
Iteration 13, loss = 0.02742802
Iteration 14, loss = 0.02737761
Iteration 15, loss = 0.02707445
Iteration 16, loss = 0.02705204
Iteration 17, loss = 0.02682275
Iteration 18, loss = 0.02662072
Iteration 19, loss = 0.02638084
Iteration 20, loss = 0.02622697
Iteration 21, loss = 0.02601328
Iteration 22, loss = 0.02598664
Iteration 23, loss = 0.02565122
Iteration 24, loss = 0.02552521
Iteration 25, loss = 0.02534948
Iteration 26, loss = 0.02510163
Iteration 27, loss = 0.02499726
Iteration 28, loss = 0.02470229
Iteration 29, loss = 0.02451417
Iteration 30, loss = 0.02446048
Iteration 31, loss = 0.02420210
Iteration 32, los

#### Below are the results of mlp with different solver and layers and it doesn't affest much.
#### Pint to be noted is that they were done indiividually and the values were mentioned for individual model result.
#### Unable to comment on how many layers will be good as the have tried multiple such combination and the results aren't
#### varying by a much larger scale.

In [17]:
'''
hidden_layer_sizes=(100,100,100,100,100)
for solver = adam 
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping at 109 iteration.
Accuracy of the mlp = 99.26%
Recall of the mlp = 49.76%
Precision of the mlp = 99.83%
F1 score of the mlp = 66.42%
Time to train 2126.64 s 

hidden_layer_sizes=(100,100,100,100,100)
for solver = sgd
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping at 74 iteration.
Accuracy of the mlp = 99.17%
Recall of the mlp = 49.64%
Precision of the mlp = 99.99%
F1 score of the mlp = 66.34%
Time to train 1065.41 s 

hidden_layer_sizes=(128,64,32,16,8)
solver='adam'
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping at 144 iter.
Accuracy of the mlp = 99.06%
Recall of the mlp = 49.91%
Precision of the mlp = 99.34%
F1 score of the mlp = 66.44%
Time to train 1783.41 s 

hidden_layer_sizes=(128,64,32,16,8)
solver='sgd'
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping at 53 iter.
Accuracy of the mlp = 99.16%
Recall of the mlp = 49.82%
Precision of the mlp = 99.61%
F1 score of the mlp = 66.42%
Time to train 620.03 s 

hidden_layer_sizes=(256,128,64,32,16)
solver='sgd'
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping at 89 iter.
Accuracy of the mlp = 99.25%
Recall of the mlp = 49.79%
Precision of the mlp = 99.76%
F1 score of the mlp = 66.43%
Time to train 1972.04 s 


solver='adam'
hidden_layer_sizes=(171,100,50)
Training loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping at iter 142.
Accuracy of the mlp = 99.06%
Recall of the mlp = 49.89%
Precision of the mlp = 99.38%
F1 score of the mlp = 66.43%
Time to train 2193.26 s 
'''

"\nhidden_layer_sizes=(100,100,100,100,100)\nfor solver = adam \nTraining loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping at 109 iteration.\nAccuracy of the mlp = 99.26%\nRecall of the mlp = 49.76%\nPrecision of the mlp = 99.83%\nF1 score of the mlp = 66.42%\nTime to train 2126.64 s \n\nhidden_layer_sizes=(100,100,100,100,100)\nfor solver = sgd\nTraining loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping at 74 iteration.\nAccuracy of the mlp = 99.17%\nRecall of the mlp = 49.64%\nPrecision of the mlp = 99.99%\nF1 score of the mlp = 66.34%\nTime to train 1065.41 s \n\nhidden_layer_sizes=(128,64,32,16,8)\nsolver='adam'\nTraining loss did not improve more than tol=0.000100 for 10 consecutive epochs. Stopping at 144 iter.\nAccuracy of the mlp = 99.06%\nRecall of the mlp = 49.91%\nPrecision of the mlp = 99.34%\nF1 score of the mlp = 66.44%\nTime to train 1783.41 s \n\nhidden_layer_sizes=(128,64,32,16,8)\nsolver='sgd'\nTraining loss

# 16. At what layer size we get the best f1-score?

## solver='adam' and hidden_layer_sizes=(171,100,50) gave good f1 score and recall value

In [18]:
mat_all=pd.DataFrame([mat_lgr, mat_lgr_sag, mat_mlp, mat_sgd, mat_ada, mat_svc, mat_xgb, mat_xscv])
mat_all.columns=['Model', 'Accuracy', 'Recall', 'Precision', 'F1-score', 'tn', 'fp', 'fn', 'tp', 'Time taken in sec']
mat_all.to_csv('model_comp.csv')
mat_all

Model  Accuracy  Recall  Precision  F1-score        tn  \
0  Logistic_regression     98.98   49.72      99.62     66.34  171245.0   
1     Log Reg with sag     99.19   49.69      99.90     66.37  171708.0   
2        MLPclassifier     99.06   49.89      99.38     66.43  170818.0   
3       SGD classifier     97.59   50.17      97.36     66.22  167323.0   
4            ada_boost     84.75   59.05      76.65     66.71  119413.0   
5                  SVC     99.17   49.67      99.91     66.36  171734.0   
6        XGBclassifier     97.71   49.96      97.89     66.16  168258.0   
7         SVC with 100     98.51   49.39      99.24     65.95   14957.0   

        fp      fn        tp  Time taken in sec  
0    638.0  2865.0  169363.0              24.42  
1    175.0  2622.0  169606.0             179.80  
2   1065.0  2160.0  170068.0            2149.34  
3   4560.0  3735.0  168493.0              21.50  
4  52470.0     0.0  172228.0             113.21  
5    149.0  2721.0  169507.0              81.46  
6   3625.0  4240.0  167988.0             274.80  
7    112.0   336.0   14595.0             249.38

# 12. If not missing a positive sample is the priority which model is best so far?

## if not missing positive sample is priority, that means the base line is of all positive, here recall has to be considered.

## from above table, it can be seen that ada_boost is having a highest recall
## Ada_boost should be used for missing less nuber of positive

# 13. If not marking negative sample as positive is top priority, which model is best so far?
# here we have to check with least percentage of false positive the models

#### first model should be selected for getting low fp. 
#### Results may vary with each iteration with a low variance( found after running multiple times.)

In [37]:
low_fp=round((mat_all.fp/(mat_all.fp+ mat_all.fn+mat_all.tp+mat_all.tn))*100,4)
mat_all['low_fp']=low_fp
low_fp=mat_all
low_fp.sort_values(by='low_fp')

Model  Accuracy  Recall  Precision  F1-score        tn  \
5                  SVC     99.17   49.67      99.91     66.36  171734.0   
1     Log Reg with sag     99.19   49.69      99.90     66.37  171708.0   
0  Logistic_regression     98.98   49.72      99.62     66.34  171245.0   
2        MLPclassifier     99.06   49.89      99.38     66.43  170818.0   
7         SVC with 100     98.51   49.39      99.24     65.95   14957.0   
6        XGBclassifier     97.71   49.96      97.89     66.16  168258.0   
3       SGD classifier     97.59   50.17      97.36     66.22  167323.0   
4            ada_boost     84.75   59.05      76.65     66.71  119413.0   

        fp      fn        tp  Time taken in sec   low_fp  
5    149.0  2721.0  169507.0              81.46   0.0433  
1    175.0  2622.0  169606.0             179.80   0.0509  
0    638.0  2865.0  169363.0              24.42   0.1854  
2   1065.0  2160.0  170068.0            2149.34   0.3095  
7    112.0   336.0   14595.0             249.38   0.3733  
6   3625.0  4240.0  167988.0             274.80   1.0534  
3   4560.0  3735.0  168493.0              21.50   1.3252  
4  52470.0     0.0  172228.0             113.21  15.2480

# Conclusion

### From all the tested methods, it was found that ADA_boost is the best model for predicting whether someone will claim the insurance or not.